azimuth/config.py

```
@@ -435,10 +449,19 @@ def load_azimuth_config(config_path: str) -> AzimuthConfig:
         If the file does not exist or the prediction file are not present.
     """
     log.info("-------------Loading Config--------------")
-    if not os.path.isfile(config_path):
-        raise EnvironmentError(f"{config_path} does not exists!")
-
-    cfg = AzimuthConfig.parse_file(config_path)
+    # Loading config from config_path if specified, or else from environment variables only.
+    cfg = AzimuthConfig.parse_file(config_path) if config_path else AzimuthConfig()
+
+    if load_config_history:
+        config_history_path = cfg.get_config_history_path()
+        try:
+            with jsonlines.open(config_history_path, mode="r") as config_history:
+                *_, last_config = config_history
+        except (FileNotFoundError, ValueError):
+            log.info("Empty or invalid config history.")
+        else:
+            log.info(f"Loading latest config from {config_history_path}.")
+            cfg = AzimuthConfig.parse_obj(last_config)
```

Should we try and except here too, and load `cfg` otherwise?

---

I'm not sure I see in which (normal) use case the latest config from `config_history.jsonl` would be invalid, so I'm not sure what the appropriate response would be, but I'm afraid it would be confusing to silently ignore the error and load another config. I am leaning towards letting the exception kill the process, just like `AzimuthConfig.parse_file(config_path)` does if the config is invalid.

To be clear, the `try`/`except` above is there to catch if the `config_history.jsonl` doesn't exist yet or is empty, which is normal on the first launch. Unfortunately, it also catches (`jsonlines` raises the same type of `ValueError` 😞) if `config_history.jsonl` contains invalid JSON (not even considering the `AzimuthConfig` schema). I don't see a (normal) use case for that to happen, so I think I would prefer to crash the process in that case too, but it's somewhat out of my control. We can revisit this later if we suspect people tend to mess up with their `config_history.jsonl` and run into this situation.

Does that make sense?

---

Yes, that makes sense. I was thinking of the use case when we introduce breaking changes when Azimuth updates, but even then, it's probably best that user deletes their cache/config history anyway .

---

Oh right, that's a good one. I'll keep thinking of what do to with breaking changes in the config. For now, let's simply be extra vigilant about not introducing breaking changes in the config, if possible.